File: pymsgmuxer/mq.py
"""Using asyncio to simulate message queue"""

import os
import asyncio
from datetime import datetime
from datetime import timedelta
from dataclasses import dataclass

class MsgNode:
    '''Message Node'''
    def __init__(self, msgid, msg):
        self.id = msgid
        self.ts = int(datetime.utcnow().timestamp() * 1000000)
        self.data = msg
        self.next = None

    def __repr__(self):
        return f'<{type(self).__name__}: {self.id}, {self.ts} => {self.data}>'

    def __eq__(self, node):
        if not node:
            return False
        return self.id == node.id and self.ts == node.ts and self.data == node.data

@dataclass
class ConsumerInfo:
    '''Consumer Information Tracking'''
    pcount: int
    lastnode: MsgNode

class MsgQueue:
    '''Message Queue'''

    begin_mark = "BeginMarker"
    end_mark = "EndMarker"

    def __init__(self, retention, cleanfreq):
        self._cinfo = {} # consumer info tracker

        self._retention = retention
        self._cleanfreq = cleanfreq

        # Using Begin and End Marker
        self._lhead = MsgNode(msgid=0, msg=MsgQueue.begin_mark)
        self._ltail = MsgNode(msgid=0, msg=MsgQueue.end_mark)
        self._lhead.next = self._ltail
        self._linsp = self._lhead

        self._seqid = 0
        self._currcnt = 0

    def put(self, ident, msg):
        '''method to put item in q'''
        msg = "[{}] {}".format(self._seqid, msg)
        node = MsgNode(msgid=ident, msg=msg)

        # The insp (insert pointer) tracks insertion point
        node.next = self._linsp.next
        self._linsp.next = node
        self._linsp = node

        self._seqid += 1
        self._currcnt += 1

        return True

    def get(self, cid):
        '''method to get item from q'''
        retmsg = None

        # Get offset dict for cid
        node = self._cinfo.get(cid, None)
        if not node:
            return retmsg

        # Get to be visiting node for cid
        node = self._cinfo[cid].lastnode
        node = node.next
        if not node or node.data == MsgQueue.end_mark:
            return retmsg

        retmsg = node.data

        # Track the consumed node
        self._cinfo[cid].lastnode = node
        self._cinfo[cid].pcount += 1
        return retmsg

    def register_consumer(self, cid):
        '''add consumer using the queue'''
        if cid in self._cinfo:
            print(f'consumer {cid} already registered, exception/error')
            return False

        # Initialize dict to store list head and count processed
        self._cinfo[cid] = ConsumerInfo(pcount=0, lastnode=self._lhead)
        return True

    def unregister_consumer(self, cid):
        '''remove consumer using the queue'''
        if cid in self._cinfo:
            del self._cinfo[cid]

    def _is_consumed(self, dnode):
        '''check if all consumers have consumed dnode'''
        cstatus = True
        cids = []
        for ckey, cdict in self._cinfo.items():
            cnode = cdict.lastnode
            if dnode == cnode:
                cstatus = False
                cids.append(ckey)
        return cstatus, cids

    def _print_status(self):
        result = "Msg Queue Status:\n"
        result += f'total = {self._seqid}, current = {self._currcnt}\n'
        #result += f'lhead: {self._lhead}, ltail: {self._ltail}\n'
        # walker = self._lhead.next
        # while walker.data != MsgQueue.end_mark:
            # result += f'  {walker!r}\n'
            # walker = walker.next
        result += f'consumer count = {len(self._cinfo.keys())}\n'
        # for ckey, cdict in self._cinfo.items():
        #     result += f'  {ckey}: count={cdict.pcount}, last-msg={cdict.lastnode!r}\n'
        print(result)

    async def cleanqueue(self):
        '''task to clean old messages'''
        procid = os.getpid()
        tstamp = datetime.utcnow().strftime("%a, %d %b %Y %H:%M:%S.%f")
        print(f'{tstamp}: Pid[{procid}] MsgQueue clean task begins...')
        while True:
            await asyncio.sleep(self._cleanfreq)
            # dbgmsg = ""
            rtime = datetime.utcnow() - timedelta(seconds=self._retention)
            rtimeint = int(rtime.timestamp() * 1000000)

            walker = self._lhead.next
            rcount = 0
            while walker.data != MsgQueue.end_mark and walker.ts < rtimeint:
                allconsume, cids = self._is_consumed(walker)
                if not allconsume:
                    # dbgmsg = f'{cids} is yet to consume {walker!r}'
                    break
                self._lhead.next = walker.next
                walker.next = None
                # print(f'Removing {walker!r}')
                walker = self._lhead.next
                self._currcnt -= 1
                rcount += 1

            # rtimestr = rtime.strftime("%a, %d %b %Y %H:%M:%S.%f")
            # if not dbgmsg:
            #     dbgmsg = f'Removing {rcount} nodes, retention time: {rtimestr}'
            # else:
            #     dbgmsg = f'Removing {rcount} nodes, retention time: {rtimestr}\n' + dbgmsg
            # print(dbgmsg)
            self._print_status()

File: gomsgmuxer/mq.go
package main

import (
	"container/list"
	"log"
	"os"
	"reflect"
	"strings"
	"sync"
	"time"
)

// ConsuserInfo tracker
type consumerInfo struct {
	pcount   int
	lastnode *list.Element
	mch      chan MsgNode
}

// Constructor for consumerInfo
func newConsumerInfo(ch chan MsgNode) *consumerInfo {
	return &consumerInfo{
		pcount:   0,
		lastnode: nil,
		mch:      ch,
	}
}

// ProducerInfo tracker
type producerInfo struct {
	pcount int
	mch    chan MsgNode
}

// Constructor for producerInfo
func newProducerInfo(ch chan MsgNode) *producerInfo {
	return &producerInfo{
		pcount: 0,
		mch:    ch,
	}
}

// MsgQueue holding the messages
type MsgQueue struct {
	sync.Mutex
	msglist              *list.List
	retention, cleanfreq int
	bulkcount            int
	seqid, currcnt       uint64
	pinfo                map[string]*producerInfo
	cinfo                map[string]*consumerInfo
}

// NewMsgQueue is constructor for MsgQueue
func NewMsgQueue(retention, cleanfreq int) *MsgQueue {
	return &MsgQueue{
		msglist:   list.New(),
		retention: retention,
		cleanfreq: cleanfreq,
		bulkcount: 10,
		seqid:     0,
		currcnt:   0,
		pinfo:     make(map[string]*producerInfo),
		cinfo:     make(map[string]*consumerInfo),
	}
}

// Put function to add message
func (m *MsgQueue) Put(pid string, mn *MsgNode) bool {
	m.msglist.PushBack(mn)
	m.seqid++
	m.currcnt++
	return true
}

// Get function to read message
func (m *MsgQueue) Get(cid string, ci *consumerInfo) (*MsgNode, bool) {
	// For cid the lastnode is not filled
	if ci.lastnode == nil {
		e := m.msglist.Front()
		if e == nil {
			// log.Printf("Consumer %s, no message in queue\n", cid)
			return nil, false
		}
		ci.lastnode = e
		return e.Value.(*MsgNode), true
	}

	ne := ci.lastnode.Next()

	// If the next node is nil, return nothing
	if ne == nil {
		// log.Printf("Consumer %s, no new messages\n", cid)
		return nil, false
	}

	// Move the CI to next node and return data
	ci.lastnode = ne
	return ne.Value.(*MsgNode), true
}

// RecvFromProducers gets messages and adds to queue
func (m *MsgQueue) RecvFromProducers() {

	procid := os.Getpid()
	tstamp := time.Now().UTC().String()
	log.Printf("%s: Pid[%d] MsgQueue.RecvFromProducers() begins...\n", tstamp, procid)

	for {
		m.Lock()

		nump := len(m.pinfo)
		// log.Printf("Num producers = %d\n", nump)
		cases := make([]reflect.SelectCase, nump+1)
		pids := make([]string, nump+1)
		i := 0
		for pid, pi := range m.pinfo {
			cases[i] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(pi.mch), Send: reflect.ValueOf(nil)}
			pids[i] = pid
			i++
		}

		if nump > 0 {
			cases[i] = reflect.SelectCase{Dir: reflect.SelectDefault, Chan: reflect.ValueOf(nil), Send: reflect.ValueOf(nil)}
			pids[i] = "default_timeout"

			for i = 0; i < m.bulkcount; i++ {
				idx, v, ok := reflect.Select(cases)
				// log.Printf("reflect.Select(): idx=%d, ok=%t\n", idx, ok)
				if ok && idx != nump {
					var mnode MsgNode
					mnode = v.Interface().(MsgNode)
					m.Put(pids[idx], &mnode)
					m.pinfo[pids[idx]].pcount++
					// log.Printf("%s: receiving pcount=%d\n", pids[idx], m.pinfo[pids[idx]].pcount)
				}
			}
		}

		m.Unlock()
		// log.Printf("Pid[%d] MsgQueue.RecvFromProducers() msg count = %d\n", procid, m.seqid)
		// time.Sleep(1 * time.Second)
	}
}

// SendToConsumers reads messages from queue and sends to consumers
func (m *MsgQueue) SendToConsumers() {

	procid := os.Getpid()
	tstamp := time.Now().UTC().String()
	log.Printf("%s: Pid[%d] MsgQueue.SendToConsumers() begins...\n", tstamp, procid)

	for {
		m.Lock()
		// From message q, send to consumer channel
		for cid, ci := range m.cinfo {
			// log.Printf("%s: checking for msg to send to consumer\n", cid)
			mnode, ok := m.Get(cid, ci)
			if ok {
				ci.pcount++
				ci.mch <- *mnode
				// log.Printf("%s: sending ccount=%d\n", cid, ci.pcount)
			}
		}

		/***
		log.Printf("Pid[%d] MsgQueue.SendToConsumers() msg count = %d\n", procid, m.seqid)
		for cid, ci := range m.cinfo {
			log.Printf("%s: sent ccount=%d\n", cid, ci.pcount)
		}
		time.Sleep(1 * time.Second)
		***/
		m.Unlock()
	}
}

// RegisterConsumer to the message queue
func (m *MsgQueue) RegisterConsumer(cid string, ch chan MsgNode) bool {
	m.Lock()
	defer m.Unlock()
	_, ok := m.cinfo[cid]
	if ok {
		log.Printf("Consumer %s already registered\n", cid)
		return false
	}
	m.cinfo[cid] = newConsumerInfo(ch)
	log.Printf("Consumer %s registration successful\n", cid)
	return true
}

// UnregisterConsumer to the message queue
func (m *MsgQueue) UnregisterConsumer(cid string) {
	m.Lock()
	defer m.Unlock()
	ci, ok := m.cinfo[cid]
	if ok {
		close(ci.mch)
		delete(m.cinfo, cid)
		log.Printf("Consumer %s unregistration successful\n", cid)
	}
}

// RegisterProducer to the message queue
func (m *MsgQueue) RegisterProducer(pid string, ch chan MsgNode) bool {
	m.Lock()
	defer m.Unlock()
	_, ok := m.pinfo[pid]
	if ok {
		log.Printf("Producer %s already registered\n", pid)
		return false
	}
	m.pinfo[pid] = newProducerInfo(ch)
	log.Printf("Producer %s registration successful\n", pid)
	return true
}

// UnregisterProducer to the message queue
func (m *MsgQueue) UnregisterProducer(pid string) {
	m.Lock()
	defer m.Unlock()
	_, ok := m.pinfo[pid]
	if ok {
		delete(m.pinfo, pid)
		log.Printf("Producer %s unregistration successful\n", pid)
	}
}

// PrintStats gives MsgQueue status
func (m *MsgQueue) PrintStats() {
	log.Println("Msg Queue Status:")
	log.Printf("total = %d, current = %d\n", m.seqid, m.currcnt)
	/***
	for e := m.msglist.Front(); e != nil; e = e.Next() {
		log.Printf("%s\n", e.Value.(*MsgNode).PrintMsgNodeStr())
	}
	***/
	log.Printf("consumer count = %d\n", len(m.cinfo))
	/***
	for cid, ci := range m.cinfo {
		s := ""
		if ci.lastnode != nil {
			s = ci.lastnode.Value.(*MsgNode).PrintMsgNodeStr()
		}
		log.Printf("%s: count=%d %s\n", cid, ci.pcount, s)
	}
	***/
}

// isConsumed checks if all consumers processed a element
func (m *MsgQueue) isConsumed(dnode *list.Element) (bool, string) {
	var cids []string
	cstatus := true
	for cid, ci := range m.cinfo {
		eq := reflect.DeepEqual(ci.lastnode, dnode)
		if eq {
			cstatus = false
			cids = append(cids, cid)
		}
	}

	cidstr := strings.Join(cids, ", ")
	return cstatus, cidstr
}

// CleanQueue cleans consumed nodes from the list
func (m *MsgQueue) CleanQueue() {
	procid := os.Getpid()
	tstamp := time.Now().UTC().String()
	log.Printf("%s: Pid[%d] MsgQueue.CleanQueue() begins...\n", tstamp, procid)

	ticker := time.NewTicker(time.Second * time.Duration(m.cleanfreq))
	for ; true; <-ticker.C {

		tnano := time.Now().UTC().UnixNano()
		rtime := tnano - int64(m.retention)*1000000000
		rcount := 0

		m.Lock()
		for {

			e := m.msglist.Front()
			if e == nil {
				break
			}

			// node timestamp not beyond retention
			if e.Value.(*MsgNode).ts >= rtime {
				break
			}

			// check if node is consumed by all consumers
			allconsume, _ := m.isConsumed(e)
			if !allconsume {
				/*** No Print in lock section
				mnode := e.Value.(*MsgNode)
				log.Printf("%s yet to consume %s", cids, mnode.PrintMsgNodeStr())
				***/
				break
			}

			m.msglist.Remove(e)
			rcount++
			m.currcnt--
		}
		m.Unlock()
		// log.Printf("Removing %d nodes, retention time %d\n", rcount, rtime)
		m.PrintStats()
	}
}
